File: Python/find-resultant-array-after-removing-anagrams.py
# Time:  O(n * l)
# Space: O(1)

import collections


# freq table
class Solution(object):
    def removeAnagrams(self, words):
        """
        :type words: List[str]
        :rtype: List[str]
        """
        result = []
        prev = None
        for x in words:
            cnt = collections.Counter(x)
            if prev and prev == cnt:
                continue
            prev = cnt
            result.append(x)
        return result


# Time:  O(n * llogl)
# Space: O(l)
import collections


# sort
class Solution2(object):
    def removeAnagrams(self, words):
        """
        :type words: List[str]
        :rtype: List[str]
        """
        result = []
        prev = None
        for x in words:
            s = sorted(x)
            if prev and prev == s:
                continue
            prev = s
            result.append(x)
        return result


# Time:  O(n * llogl)
# Space: O(l)
import collections


# sort
class Solution3(object):
    def removeAnagrams(self, words):
        """
        :type words: List[str]
        :rtype: List[str]
        """
        return [words[i] for i in xrange(len(words)) if i == 0 or sorted(words[i-1]) != sorted(words[i])]

File: C++/find-resultant-array-after-removing-anagrams.cpp
// Time:  O(n * l)
// Space: O(1)

// freq table
class Solution {
public:
    vector<string> removeAnagrams(vector<string>& words) {
        vector<string> result;
        vector<int> prev;
        for (const auto& x : words) {
            vector<int> cnt(26);
            for (const auto& c : x) {
                ++cnt[c - 'a'];
            }
            if (!empty(prev) && prev == cnt) {
                continue;
            }
            prev = move(cnt);
            result.emplace_back(x);
        }
        return result;
    }
};

// Time:  O(n * llogl)
// Space: O(l)
// sort
class Solution2 {
public:
    vector<string> removeAnagrams(vector<string>& words) {
        vector<string> result;
        string prev;
        for (const auto& x : words) {
            string s(x);
            sort(begin(s), end(s));
            if (!empty(prev) && prev == s) {
                continue;
            }
            prev = move(s);
            result.emplace_back(x);
        }
        return result;
    }
};
